### PR TITLE
feat: add KRH.delete, KRH.reconcile, and supporting helpers/tests

### PR DESCRIPTION
**Pull request by ca-scribner**
Added here are features tracking, managing, and removing resources from previous executions.  

This PR includes:
* new inputs `labels` and `child_resource_types` for KRH, to help with managing resources deployed previously
* KRH.get_deployed_resources(), which finds previously deployed resources in a cluster that are "owned" by this instantiation of KRH
* KRH.delete(), which deletes previously deployed resources
* KRH.reconcile(), which reconciles (deletes, modifies, and creates) previous resources to match the current desired state
* helpers and tests for the above features
* new integration tests for KRH (apply, reconcile, and delete)

### CODE REVIEW

File: src/charmed_kubeflow_chisme/kubernetes/_kubernetes_resource_handler.py [hunk 1]
```diff
@@ -289,3 +412,95 @@ def _charm_status_given_resource_status(
...
+    Args:
+        resources: a list of Lightkube resources to validate
+        allowed_resource_types: a list of Lightkube resource classes to validate against
+    """
+    resource_types = _get_resource_classes_in_manifests(resources)
+    for resource_type in resource_types:
+        if resource_type not in allowed_resource_types:
+            raise ValueError(
+                f"Resource type {resource_type} not in allowed resource types"
+                f" '{allowed_resource_types}'"
```
Comment (i-chvets) [inline src/charmed_kubeflow_chisme/kubernetes/_kubernetes_resource_handler.py:505]: Can we also have it in logs?

File: tests/integration/test_kubernetes.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+    try:
+        lightkube_client.delete(Namespace, namespace_name)
+    except ApiError as err:
+        if err.code == 404:
+            return
+        raise err
+
+
+def test_KubernetesResourceHandler_apply(namespace):  # noqa: N802
```
Comment (i-chvets) [inline tests/integration/test_kubernetes.py:48]: Doc string.

File: tests/integration/test_kubernetes.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    lightkube_client.get(Namespace, additional_namespace_name)
+
+    # Reconcile away the Services by updating the template files to remove the services and
+    # namespace
+    krh.template_files = [data_dir / "pods1.j2"]
+    krh.reconcile()
+
+    # Give a few seconds for everything to be deleted
+    time.sleep(10)
+
```
Comment (i-chvets) [inline tests/integration/test_kubernetes.py:81]: I don't like this relying on sleep. What else can we do? Use tenacity?

File: tests/integration/test_kubernetes.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    with pytest.raises(ApiError) as err:
+        lightkube_client.get(Namespace, additional_namespace_name)
+    assert err.value.status.code == 404
+
+    # Delete everything we created
+    krh.delete()
+
+    # Give a few seconds for everything to be deleted
+    time.sleep(10)
+
```
Comment (i-chvets) [inline tests/integration/test_kubernetes.py:101]: Same as above.

File: tests/integration/test_kubernetes.py (hunk 2 above)
Comment (i-chvets) [inline tests/integration/test_kubernetes.py:48]: Doc string.

File: src/charmed_kubeflow_chisme/kubernetes/_kubernetes_resource_handler.py (hunk 1 above)
Comment (ca-scribner) [reply src/charmed_kubeflow_chisme/kubernetes/_kubernetes_resource_handler.py:505]: Should this helper write to a logger, or should the charm calling this log the error? I'm never sure where to put these things.

If we put a log statement here, then it'll always log the error even if whoever called this function (eg: a charm) caught it and handled it. Would that be confusing?

File: tests/integration/test_kubernetes.py (hunk 2 above)
Comment (ca-scribner) [reply tests/integration/test_kubernetes.py:48]: Added

File: tests/integration/test_kubernetes.py (hunk 3 above)
Comment (ca-scribner) [reply tests/integration/test_kubernetes.py]: Done

File: tests/integration/test_kubernetes.py (hunk 4 above)
Comment (ca-scribner) [reply tests/integration/test_kubernetes.py]: Done

**Review by i-chvets — APPROVED** [review]
Approved.

**Pull request merged**